_posts/2023-04-21-tensorflow-mnist-classification.adoc

```diff
@@ -0,0 +1,405 @@
+
+[source,python]
+----
+model = tf.keras.models.Sequential([
+    tf.keras.layers.Dense(128, activation='relu'),
+    tf.keras.layers.Dense(10)
+])
+----
+
+This model contains only two layers.
+Although this model is really simple, it still does a pretty good job in recognition of handwritten digits.
+Probably more interesting than the model itself is how easy it is to write a mode in TensorFlow (or actually https://keras.io/[Keras], but it's now part of TensorFlow).
+
+Similarly easy is to define model optimizer and the loss function:
+
+[source,python]
+----
+model.compile(
+    optimizer=tf.keras.optimizers.Adam(0.001),
+    loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
+    metrics=[tf.keras.metrics.SparseCategoricalAccuracy()],
+)
+----
+
+It's outside of the scope of this post to explain what these functions are and you can check almost any machine learning online course or textbook on this topic for a detailed explanation.
+
+Once we have out model ready, we can train it on the train dataset prepared in previous section:
+
+[source,python]
+----
+model.fit(train_ds,epochs=MAX_EPOCHS)
+----
+
+This step may take quite some time to finish.
+However, once finished, our model is ready to recognize handwritten digits!
+
+=== Streaming the data into the model
+
+Let's see how good our model is in digit recognition.
+But as our main goal here is to explore the means how to ingest data into TensorFlow, we will start model evaluation on empty (or more accurately even non-existing) Kafka topic and see if we will be able to evaluate the data on the fly as they will pop-up first in the database and then in the corresponding Kafka topic.
+For this purpose we can use one of the above-mentioned streaming classes:
+
+[source,python]
+----
+test_ds = tfio.experimental.streaming.KafkaGroupIODataset(
+    topics=[KAFKA_TEST_TOPIC],
+    group_id=KAFKA_CONSUMER_GROUP,
+    servers=KAFKA_SERVERS,
+    stream_timeout=9000,
+    configuration=[
+        "session.timeout.ms=10000",
+        "max.poll.interval.ms=10000",
+        "auto.offset.reset=earliest"
+    ],
+)
+----
+
+Again, arguments are mostly self-explanatory.
+Two things may need further explanation: `stream_timeout` and `configuration` parameters.
+`stream_timeout` determines interval of inactivity (in milliseconds) after which the streaming would terminate.
+`configuration` is https://github.com/confluentinc/librdkafka/blob/master/CONFIGURATION.md[librdkafka configuration].
+It's basically a configuration of Kafka client and you should configure at least session timeout (`session.timeout.ms`) and its poll interval (`max.poll.interval.ms`).
+The values of these parameters should be higher than the value of `stream_timeout`.
+
+The dataset provided by this loader is slightly different - instead of providing a single record containing message and its key, we get key and message already split.
+Therefore, we have to define slightly modified map function with two arguments:
```

Comment:
```suggestion
The dataset this loader provides is slightly different - instead of providing a single record containing the message and its key, we get the key and message already split.
Therefore, we have to define a slightly modified map function with two arguments:
```